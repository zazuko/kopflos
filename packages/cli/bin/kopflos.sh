#!/usr/bin/env sh

# find JS entrypoint
kopflos=$(node -e "console.log(require.resolve('kopflos/bin/index.js'))" 2> /dev/null)

# if tsx or ts-node exists in path, use them
if command -v tsx > /dev/null 2>&1
then
  node --import tsx --no-warnings "$kopflos" "$@"
elif command -v ts-node > /dev/null 2>&1
then
  node --loader ts-node/esm/transpile-only --no-warnings "$kopflos" "$@"
else
  node "$kopflos" "$@"
fi
