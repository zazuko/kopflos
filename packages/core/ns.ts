import rdf from '@zazuko/env-node'

type Properties = 'api' | 'resourceLoader' | 'handler' | 'method' | 'config'
type Classes = 'Config' | 'Api' | 'ResourceShape' | 'Handler'
type Shorthands = 'DescribeLoader' | 'OwnGraphLoader'

export interface KopflosTerms extends Record<Properties | Classes | Shorthands, never>{
}

export const kl = rdf.namespace<keyof KopflosTerms>('https://kopflos.described.at/')
