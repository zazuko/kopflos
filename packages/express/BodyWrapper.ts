import { Readable } from 'node:stream'
import type { Request } from 'express'
import type { NamedNode } from '@rdfjs/types'
import type { Body } from '@kopflos-cms/core'
import type { Environment } from '@rdfjs/environment/Environment.js'
import type ClownfaceFactory from 'clownface/Factory.js'
import type { Dataset } from '@zazuko/env/lib/DatasetExt.js'
import type { DatasetFactoryExt } from '@zazuko/env/lib/DatasetFactoryExt.js'

export class BodyWrapper implements Body {
  private _dataset?: Promise<Dataset>

  constructor(private readonly env: Environment<DatasetFactoryExt | ClownfaceFactory>, private readonly term: NamedNode, private readonly req: Readable & Pick<Request, 'quadStream'>) {

  }

  get dataset(): Promise<Dataset> {
    if (!this._dataset) {
      this._dataset = this.env.dataset().import(this.quadStream)
    }
    return this._dataset
  }

  get quadStream() {
    return this.req.quadStream!()
  }

  async pointer() {
    return this.env.clownface({
      dataset: await this.dataset,
      term: this.term,
    })
  }

  get raw() {
    return Readable.toWeb(this.req)
  }
}
