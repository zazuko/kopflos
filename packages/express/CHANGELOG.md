# @kopflos-cms/express

## 0.1.0

### Minor Changes

- fd489b3: First version
- 1fcb2c0: The main export now returns `Promise<{ middleware: RequestHandler; instance: Kopflos }>`
- fd489b3: Added express-only middleware hooks

### Patch Changes

- 67ba100: Added CHANGELOG.md to build
- 3e30f38: Support for direct stream, using Web Streams
- 67ba100: Plugin dir wasn't packaged
- b94672f: Body is now always present in handler argument. Use `this.body.isRdf` to check if it can be parsed as RDF according to the Accept header. `this.body.raw` now returns the `IncomingMessage` object to allow parsing bodies using [`co-body`](https://npm.im/co-body) or similar.
- 3a227e5: First version
- 14705ba: Adding support for accessing request body
- fd489b3: Added plugin with `onStart` hook
- facbd83: build(deps): bump express from 4.19.2 to 4.20.0
- 730ecc2: Ensure that middlewares from multiple plugins are registered in order
- 1b4a7fc: After first request, the server would stop responding
- Updated dependencies [730ecc2]
- Updated dependencies [67ba100]
- Updated dependencies [619d7f2]
- Updated dependencies [3e30f38]
- Updated dependencies [b691125]
- Updated dependencies [77cfe87]
- Updated dependencies [1fcb2c0]
- Updated dependencies [14705ba]
- Updated dependencies [fe2793d]
- Updated dependencies [b72086f]
- Updated dependencies [673f9a2]
- Updated dependencies [3e30f38]
- Updated dependencies [2b107ef]
- Updated dependencies [3e30f38]
- Updated dependencies [67ba100]
- Updated dependencies [13f029c]
- Updated dependencies [b94672f]
- Updated dependencies [528bfd2]
- Updated dependencies [1b8a2bc]
- Updated dependencies [0b960fa]
- Updated dependencies [67ba100]
- Updated dependencies [a72254b]
- Updated dependencies [a11dda5]
- Updated dependencies [14705ba]
- Updated dependencies [2ee7a70]
- Updated dependencies [fd489b3]
- Updated dependencies [67bd393]
- Updated dependencies [be93e5a]
- Updated dependencies [3e30f38]
- Updated dependencies [730ecc2]
  - @kopflos-cms/core@0.3.0
  - @kopflos-cms/logger@0.1.0

## 0.1.0-beta.6

### Patch Changes

- 67ba100: Added CHANGELOG.md to build
- 67ba100: Plugin dir wasn't packaged
- Updated dependencies [67ba100]
- Updated dependencies [67ba100]
- Updated dependencies [67ba100]
  - @kopflos-cms/core@0.3.0-beta.10
  - @kopflos-cms/logger@0.1.0-beta.1

## 0.1.0-beta.5

### Minor Changes

- fd489b3: First version
- 1fcb2c0: The main export now returns `Promise<{ middleware: RequestHandler; instance: Kopflos }>`
- fd489b3: Added express-only middleware hooks

### Patch Changes

- 3e30f38: Support for direct stream, using Web Streams
- fd489b3: Added plugin with `onStart` hook
- 730ecc2: Ensure that middlewares from multiple plugins are registered in order
- Updated dependencies [730ecc2]
- Updated dependencies [3e30f38]
- Updated dependencies [1fcb2c0]
- Updated dependencies [3e30f38]
- Updated dependencies [3e30f38]
- Updated dependencies [a72254b]
- Updated dependencies [fd489b3]
- Updated dependencies [be93e5a]
- Updated dependencies [3e30f38]
- Updated dependencies [730ecc2]
  - @kopflos-cms/core@0.3.0-beta.9
  - @kopflos-cms/logger@0.1.0-beta.0

## 0.0.1-beta.4

### Patch Changes

- facbd83: build(deps): bump express from 4.19.2 to 4.20.0
- Updated dependencies [b72086f]
  - @kopflos-cms/core@0.3.0-beta.8

## 0.0.1-beta.3

### Patch Changes

- b94672f: Body is now always present in handler argument. Use `this.body.isRdf` to check if it can be parsed as RDF according to the Accept header. `this.body.raw` now returns the `IncomingMessage` object to allow parsing bodies using [`co-body`](https://npm.im/co-body) or similar.
- Updated dependencies [2b107ef]
- Updated dependencies [b94672f]
- Updated dependencies [67bd393]
  - @kopflos-cms/core@0.3.0-beta.6

## 0.0.1-beta.2

### Patch Changes

- 14705ba: Adding support for accessing request body
- Updated dependencies [b691125]
- Updated dependencies [14705ba]
- Updated dependencies [14705ba]
- Updated dependencies [2ee7a70]
  - @kopflos-cms/core@0.3.0-beta.5

## 0.0.1-beta.1

### Patch Changes

- 1b4a7fc: After first request, the server would stop responding
- Updated dependencies [673f9a2]
  - @kopflos-cms/core@0.3.0-beta.2

## 0.0.1-beta.0

### Patch Changes

- 3a227e5: First version
- Updated dependencies [77cfe87]
- Updated dependencies [528bfd2]
  - @kopflos-cms/core@0.3.0-beta.0
