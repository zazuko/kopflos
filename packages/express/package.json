{
  "name": "@kopflos-cms/express",
  "version": "0.0.1-beta.0",
  "main": "index.js",
  "license": "MIT",
  "type": "module",
  "scripts": {
    "test": "mocha",
    "build": "tsc",
    "prepack": "npm run build"
  },
  "exports": {
    ".": "./index.js"
  },
  "files": [
    "*.js",
    "*.d.ts",
    "lib/**/*.js",
    "lib/**/*.d.ts"
  ],
  "dependencies": {
    "@kopflos-cms/core": "^0.3.0-beta.0",
    "@rdfjs/express-handler": "^2.0.2",
    "@zazuko/env-node": "^2.1.3",
    "absolute-url": "^2.0.0",
    "express": "^4.19.2",
    "middleware-async": "^1.4.0",
    "onetime": "^7.0.0",
    "ts-pattern": "^5.3.1"
  },
  "devDependencies": {
    "@types/absolute-url": "^2.0.0",
    "@types/express": "^4.17.21",
    "@types/rdfjs__express-handler": "^2.0.6",
    "@types/supertest": "^6.0.2",
    "chai": "^4.5.0",
    "mocha-chai-rdf": "*",
    "supertest": "^7.0.0"
  },
  "author": "Zazuko GmbH",
  "repository": {
    "type": "git",
    "url": "git://github.com/zazuko/kopflos.git",
    "directory": "packages/express"
  },
  "bugs": {
    "url": "https://github.com/zazuko/kopflos/issues"
  },
  "homepage": "https://github.com/zazuko/kopflos",
  "mocha": {
    "extension": [
      "ts"
    ],
    "spec": "test/**/*.test.ts",
    "loader": "ts-node/esm",
    "require": "../../mocha-setup.js"
  }
}
