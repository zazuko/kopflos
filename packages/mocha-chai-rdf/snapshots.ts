/* eslint-disable @typescript-eslint/no-explicit-any */
import chai, { expect } from 'chai'
import { jestSnapshotPlugin } from 'mocha-chai-jest-snapshot'
import rdf from '@zazuko/env-node'

declare global {
  /* eslint-disable @typescript-eslint/no-namespace */
  namespace Chai {
    interface Assertion {
      canonical: Assertion
    }
  }
}

chai.use(jestSnapshotPlugin())

chai.use((_chai, utils) => {
  const toMatchSnapshot = (chai.Assertion.prototype as any).toMatchSnapshot

  utils.addMethod(chai.Assertion.prototype, 'toMatchSnapshot', function (this: Chai.Assertion, ...args: any[]) {
    const obj = (this as any)._obj

    if (utils.flag(this, 'dataset-canonical')) {
      // Custom behavior for dataset
      expect(rdf.dataset.toCanonical(obj)).toMatchSnapshot(...args)
    } else {
      // Original behavior
      toMatchSnapshot.apply(this, args)
    }
  })

  utils.addProperty(chai.Assertion.prototype, 'canonical', function (this: Chai.Assertion) {
    utils.flag(this, 'dataset-canonical', true)
  })
})
