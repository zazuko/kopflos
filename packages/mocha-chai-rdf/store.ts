/* eslint-disable no-console,camelcase */
import * as url from 'node:url'
import assert from 'node:assert'
import * as Oxigraph from 'oxigraph'
import type { NamespaceBuilder } from '@rdfjs/namespace'
import rdf from '@zazuko/env-node'
import type { DatasetCore, NamedNode, Quad_Graph } from '@rdfjs/types'
import type { AnyPointer } from 'clownface'
import type { Dataset } from '@zazuko/env/lib/Dataset.js'

declare module 'mocha' {
  interface Context {
    dataset: DatasetCore
    graph: AnyPointer
    store: Oxigraph.Store
  }
}

interface DatasetSourceOptions {
  format?: 'trig' | 'nq'
  loadAll?: boolean
  includeDefaultGraph?: boolean
}

interface GraphSourceOptions {
  format?: 'ttl' | 'nt'
}

type Options = (DatasetSourceOptions | GraphSourceOptions) & {
  baseIri?: string | NamespaceBuilder
}

export function createStore(base: string, options: Options = {}) {
  const format = options.format ?? 'ttl'
  let loadAll = true
  let includeDefaultGraph = false
  if (options.format === 'trig' || options.format === 'nq') {
    loadAll = options.loadAll || false
    includeDefaultGraph = options.includeDefaultGraph || false
  }

  return async function (this: Mocha.Context) {
    const store = new Oxigraph.Store()

    const path = url.fileURLToPath(new url.URL(`${base}.${format}`))

    let dataset: Dataset = await rdf.dataset().import(rdf.fromFile(path))

    let graph: Quad_Graph | undefined
    if (this.currentTest && !loadAll) {
      graph = testGraph(this.currentTest)
      dataset = dataset
        .filter((quad) => {
          return quad.graph.equals(graph) || (includeDefaultGraph && quad.graph.equals(rdf.defaultGraph()))
        })
        .map(function toDefaultGraph(quad) {
          return rdf.quad(quad.subject, quad.predicate, quad.object)
        })
    }

    for (const quad of dataset) {
      store.add(quad)
    }

    function assertNotEmpty() {
      if (loadAll) {
        return assert(dataset.size > 0, 'Test data not found')
      }
      if (graph && !graph.equals(rdf.defaultGraph())) {
        return assert(
          dataset.size > 0,
          `Test data not found in GRAPH <${graph.value}>`,
        )
      }

      return assert(
        dataset.match(null, null, null, graph).size > 0,
        `Test data not found in GRAPH <${graph?.value}>`,
      )
    }

    Object.defineProperty(this, 'dataset', {
      get() {
        assertNotEmpty()
        return dataset
      },
      configurable: true,
    })

    Object.defineProperty(this, 'graph', {
      get() {
        assertNotEmpty()
        return rdf.clownface({ dataset })
      },
      configurable: true,
    })
    Object.defineProperty(this, 'store', {
      get() {
        assertNotEmpty()
        return store
      },
      configurable: true,
    })

    // hack to remove the graph after the test just at the right moment
    const cleanup = this.test?.title.includes('before all') ? after : afterEach
    cleanup(() => {
      /* eslint-disable @typescript-eslint/ban-ts-comment */
      // @ts-ignore
      delete this.dataset
      // @ts-ignore
      delete this.graph
      // @ts-ignore
      delete this.store
    })
  }
}

function testGraph(test: Mocha.Test): NamedNode {
  return rdf.namedNode(encodeURI(test.titlePath().slice(1, -1).map(removeSpaces).join('/')))
}

function removeSpaces(arg: string) {
  return arg.replaceAll(' ', '-')
}
