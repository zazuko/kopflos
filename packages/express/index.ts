import type { RequestHandler } from 'express'
import { Router } from 'express'
import type { KopflosConfig, Query } from '@kopflos-cms/core'
import Kopflos from '@kopflos-cms/core'
import absolutUrl from 'absolute-url'
import rdfHandler from '@rdfjs/express-handler'
import factory from '@zazuko/env-node'
import onetime from 'onetime'
import { match, P } from 'ts-pattern'
import asyncMiddleware from 'middleware-async'
import { loadPlugins } from '@kopflos-cms/core/plugins.js' // eslint-disable-line import/no-unresolved
import { BodyWrapper } from './BodyWrapper.js'

declare module 'express-serve-static-core' {
  interface Request {
    iri: string
  }
}

interface MiddlewareHook {
  (host: Router, instance: Kopflos): Promise<void> | void
}

declare module '@kopflos-cms/core' {
  interface KopflosPlugin {
    beforeMiddleware?: MiddlewareHook
    afterMiddleware?: MiddlewareHook
  }
}

export default async (options: KopflosConfig): Promise<{ middleware: RequestHandler; instance: Kopflos }> => {
  const kopflos = new Kopflos(options, {
    plugins: await loadPlugins(options.plugins),
  })

  const loadApiGraphs = onetime(async (graphs: Required<KopflosConfig>['apiGraphs']) => {
    await Kopflos.fromGraphs(kopflos, ...graphs)
  })

  const router = Router()

  await registerMiddlewares(router, kopflos, kopflos.plugins.map(plugin => plugin.beforeMiddleware))

  router
    .use((req, res, next) => {
      if (!options.apiGraphs) {
        return next(new Error('No API graphs configured. In future release it will be possible to select graphs dynamically.'))
      }

      loadApiGraphs(options.apiGraphs).then(next).catch(next)
    })
    .use((req, res, next) => {
      const fullUrl = absolutUrl(req)
      fullUrl.search = ''
      req.iri = fullUrl.toString()
      next()
    })
    .use(rdfHandler({
      factory,
      baseIriFromRequest: (req) => req.iri,
    }))
    .use(asyncMiddleware(async (req, res, next) => {
      const result = await kopflos.handleRequest({
        method: req.method,
        headers: req.headers,
        iri: kopflos.env.namedNode(req.iri),
        body: new BodyWrapper(kopflos.env, kopflos.env.namedNode(req.iri), req),
        query: req.query as unknown as Query,
      })

      if (result.status === 404) {
        return next()
      }

      if (result.status) {
        res.status(result.status)
      }
      res.set(result.headers)
      await match(result.body)
        .with(P.string, body => res.send(body))
        .with(P.nullish, () => res.end())
        .with(P.instanceOf(Error), error => next(error))
        .with({ size: P.number }, (dataset) => res.dataset(dataset))
        .with({ terms: P.array() }, ({ dataset }) => res.dataset(dataset))
        .with({ read: P.any }, stream => res.quadStream(stream))
        .otherwise((stream) => res.send(stream))
    }))

  await registerMiddlewares(router, kopflos, kopflos.plugins.map(plugin => plugin.afterMiddleware))

  return {
    middleware: router,
    instance: kopflos,
  }
}

async function registerMiddlewares(router: Router, kopflos: Kopflos, hooks: Array<MiddlewareHook | undefined>) {
  for (const hook of hooks) {
    await hook?.(router, kopflos)
  }
}
