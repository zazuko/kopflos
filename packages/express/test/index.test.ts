import * as url from 'node:url'
import type { Express } from 'express'
import express from 'express'
import request from 'supertest'
import { createStore } from 'mocha-chai-rdf/store.js'
import snapshots from 'mocha-chai-rdf/snapshots.js'
import { expect, use } from 'chai'
import { ntriples, turtle } from '@tpluscode/rdf-string'
import rdf from '@zazuko/env-node'
import { ex } from '../../testing-helpers/ns.js'
import kopflos from '../index.js'
import inMemoryClients from '../../testing-helpers/in-memory-clients.js'

describe('@kopflos-cms/express', () => {
  use(snapshots)

  let app: Express

  beforeEach(createStore(import.meta.url, { format: 'trig', loadAll: true, sliceTestPath: [2, -1] }))

  describe('configured api graphs', () => {
    beforeEach(function () {
      app = express()
        .set('trust proxy', true)
        .use(kopflos({
          codeBase: url.fileURLToPath(new URL('.', import.meta.url)),
          sparql: {
            default: inMemoryClients(this.rdf),
          },
          apiGraphs: [ex.api],
        }))
    })

    it('should call next when no resource shape is found', async () => {
      app.use((req, res) => {
        res.send('not found')
      })

      await request(app)
        .get('/not-found')
        .expect(200, 'not found')
    })

    it('must not stall after first request', async () => {
      await request(app)
        .get('/not-found')
        .expect(404)

      await request(app)
        .get('/not-found')
        .expect(404)
    })

    context('request to resource without explicit handler', () => {
      it('should should return Core representation', async () => {
        const response = await request(app)
          .get('/no-handler')
          .set('host', 'example.org')
          .expect(200)

        expect(response.body).toMatchSnapshot()
      })

      for (const method of ['post', 'put'] as const) {
        it('should should return 405 to ' + method.toUpperCase(), async () => {
          await request(app)[method]('/no-handler')
            .set('host', 'example.org')
            .expect(405)
        })
      }

      it('should support content negotiation', async () => {
        const response = await request(app)
          .get('/no-handler')
          .set('host', 'example.org')
          .set('accept', 'text/turtle')
          .expect('Content-Type', /text\/turtle/)
          .expect(200)

        expect(response.text).toMatchSnapshot()
      })
    })

    it('should respond to HEAD requests', async () => {
      const response = await request(app)
        .head('/no-handler')
        .set('host', 'example.org')
        .expect(200)

      expect(response.body).to.be.empty
    })

    context('request to resource with GET handler', () => {
      it('should should return Core representation', async () => {
        const response = await request(app)
          .get('/with-handler')
          .set('host', 'example.org')
          .expect(200)

        expect(response.body).toMatchSnapshot()
      })

      it('should respond to HEAD requests', async () => {
        const response = await request(app)
          .head('/with-handler')
          .set('host', 'example.org')
          .expect(200)

        expect(response.body).to.be.empty
      })

      it('should support content negotiation', async () => {
        const response = await request(app)
          .get('/with-handler')
          .set('host', 'example.org')
          .set('accept', 'text/turtle')
          .expect('Content-Type', /text\/turtle/)
          .expect(200)

        expect(response.text).toMatchSnapshot()
      })

      it('should forward response headers', async () => {
        await request(app)
          .head('/with-handler')
          .set('host', 'example.org')
          .expect(200)
          .expect('x-handler', 'get-with-handler.js')
      })
    })

    context('request with body', () => {
      it('should be undefined on GET requests', async () => {
        const { body } = await request(app)
          .get('/body-handler')
          .send(turtle`<> a ${ex.Foo} .`.toString())
          .set('host', 'example.org')
          .set('accept', 'application/n-triples')
          .expect(200)

        expect(body).to.be.empty
      })

      for (const type of ['stream', 'dataset', 'pointer']) {
        it(`body can be accessed as ${type}`, async () => {
          const response = await request(app)
            .post('/body-handler')
            .query({ type })
            .set('content-type', 'text/turtle')
            .send(turtle`<> a ${ex.Foo} .`.toString())
            .set('host', 'example.org')
            .set('accept', 'application/n-triples')
            .expect(200)

          expect(response.text).to.eq(ntriples`<http://example.org/body-handler> ${rdf.ns.rdf.type} ${ex.Foo} .\n`.toString())
        })
      }

      it('can parse body directly', async () => {
        const { body } = await request(app)
          .get('/body-handler')
          .query({ type: 'json' })
          .send({ foo: 'baz' })
          .set('content-type', 'application/json')
          .set('host', 'example.org')
          .expect(200)

        expect(body).to.deep.eq({
          bar: 'baz',
        })
      })
    })
  })

  describe('automatic API graph', () => {
    beforeEach(function () {
      app = express()
        .set('trust proxy', true)
        .use(kopflos({
          codeBase: url.fileURLToPath(new URL('.', import.meta.url)),
          sparql: {
            default: inMemoryClients(this.rdf),
          },
        }))
    })

    it('throws an error when no API graphs are configured', async () => {
      await request(app)
        .get('/')
        .expect(500, /No API graphs configured/)
    })
  })
})
