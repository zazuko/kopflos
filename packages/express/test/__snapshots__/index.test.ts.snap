// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@kopflos-cms/express configured api graphs request to resource with GET handler should should return Core representation 1`] = `
Array [
  Object {
    "@id": "http://example.org/with-handler",
    "http://schema.org/name": "from handler code",
  },
]
`;

exports[`@kopflos-cms/express configured api graphs request to resource with GET handler should support content negotiation 1`] = `
"<http://example.org/with-handler> <http://schema.org/name> \\"from handler code\\" .
"
`;

exports[`@kopflos-cms/express configured api graphs request to resource without explicit handler should should return Core representation 1`] = `
Array [
  Object {
    "@id": "http://example.org/no-handler",
    "http://www.w3.org/2000/01/rdf-schema#label": "No handler",
  },
  Object {
    "@id": "http://example.org/no-handler",
    "@type": "http://example.org/NoHandler",
  },
]
`;

exports[`@kopflos-cms/express configured api graphs request to resource without explicit handler should support content negotiation 1`] = `
"<http://example.org/no-handler> <http://www.w3.org/2000/01/rdf-schema#label> \\"No handler\\" .
<http://example.org/no-handler> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://example.org/NoHandler> .
"
`;

exports[`@kopflos-cms/express request to resource with GET handler should should return Core representation 1`] = `
Array [
  Object {
    "@id": "http://example.org/with-handler",
    "http://schema.org/name": "from handler code",
  },
]
`;

exports[`@kopflos-cms/express request to resource with GET handler should support content negotiation 1`] = `
"<http://example.org/with-handler> <http://schema.org/name> \\"from handler code\\" .
"
`;

exports[`@kopflos-cms/express request to resource without explicit handler should should return Core representation 1`] = `
Array [
  Object {
    "@id": "http://example.org/no-handler",
    "http://www.w3.org/2000/01/rdf-schema#label": "No handler",
  },
  Object {
    "@id": "http://example.org/no-handler",
    "@type": "http://example.org/NoHandler",
  },
]
`;

exports[`@kopflos-cms/express request to resource without explicit handler should support content negotiation 1`] = `
"<http://example.org/no-handler> <http://www.w3.org/2000/01/rdf-schema#label> \\"No handler\\" .
<http://example.org/no-handler> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://example.org/NoHandler> .
"
`;
